controllers/etcd_custodian_controller.go

```diff
@@ -131,43 +133,47 @@ func (ec *EtcdCustodian) updateEtcdStatus(ctx context.Context, logger logr.Logge
 	)
 
 	return kutil.TryUpdateStatus(ctx, retry.DefaultBackoff, ec.Client, etcd, func() error {
-		etcd.Status.Etcd = &druidv1alpha1.CrossVersionObjectReference{
-			APIVersion: sts.APIVersion,
-			Kind:       sts.Kind,
-			Name:       sts.Name,
-		}
-
 		etcd.Status.Conditions = conditions
 		etcd.Status.Members = members
 
-		ready := CheckStatefulSet(etcd, sts) == nil
-
-		// To be changed once we have multiple replicas.
-		etcd.Status.CurrentReplicas = sts.Status.CurrentReplicas
-		etcd.Status.ReadyReplicas = sts.Status.ReadyReplicas
-		etcd.Status.UpdatedReplicas = sts.Status.UpdatedReplicas
-		etcd.Status.Ready = &ready
-		logger.Info(fmt.Sprintf("ETCD status updated for statefulset current replicas: %v, ready replicas: %v, updated replicas: %v", sts.Status.CurrentReplicas, sts.Status.ReadyReplicas, sts.Status.UpdatedReplicas))
-		return nil
-	})
-}
+		// Bootstrap is a special case which is handled by the etcd controller.
+		if !inBootstrap(etcd) && len(members) != 0 {
+			etcd.Status.ClusterSize = pointer.Int32Ptr(int32(len(members)))
+		}
```

Comment:
Strictly speaking, this is incorrect. Only the initial bootstrapping is triggered by the etcd controller. Later bootstrapping due to [quorum failure](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/multi-node/README.md#10-quorum-lost-with-a-majority-of-members-notready) is to be [done by the custodian controller](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/multi-node/README.md#recommended-action-9). But this change could be done in https://github.com/gardener/etcd-druid/issues/158.

Reply:
I added this comment as this reflects the current state.

Tbh, I'm not convinced yet that the Custodian controller performs self healing in the form of re-bootstrapping because the Etcd controller can in the meantime still reconcile `etcd` resources which can lead to races. In order to avoid a race, you need to introduce some kind of lock which is also not too nice, IMO.

Maybe I didn't check all cases in detail but at the moment I think an alternative would be easier to implement:
1. Custodian sees that cluster is not healthy and self healing is required.
2. Custodian triggers a reconciliation for the affected `etcd` resource.
3. Etcd controller kicks in and takes necessary action to bring cluster up again (=> desired state). In the meantime any changes to the very same `etcd` resource cannot lead to a race.

We don't have to decide in this PR of course, but still wanted to give an alternative approach to think about.

Reply:
> In order to avoid a race, you need to introduce some kind of lock which is also not too nice, IMO.

Even if only the etcd controller is responsible for bootstrapping non-quorate clusters, some sort of race avoidance is still required because the `Etcd` spec might get updated while bootstrapping is still on-going at which case it should either wait until the on-going bootstrapping completed before applying the latest spec changes or should abort the current bootstrapping (which might unnecessarily prolong the downtime).

> 2. Custodian triggers a reconciliation for the affected etcd resource.

This cannot be the regular etcd controller reconciliation with which might apply `Etcd` resource spec changes along with triggering bootstrapping. As a part of the contract of a gardener extension, `etcd-druid` not support apply `Etcd` resource spec changes unless `reconcile` `operation` annotation is present.

On the other hand, it makes no sense for the `etcd-druid` to wait for the next shoot reconciliation to fix a non-quorate ETCD cluster.

But let's take this discussion out of this PR.